maintenance/githubutil/githubutil.go

```diff
@@ -0,0 +1,195 @@
+	delay := c.retryInitialBackoff * time.Duration(math.Exp2(float64(retryCount)))
+	if delay > maxDelay {
+		delay = maxDelay
+	}
+	time.Sleep(delay)
+}
+
+func (c *Client) limitRate(r *github.Rate) {
+	if r.Remaining <= c.tokenReserve {
+		sleepDuration := time.Until(r.Reset.Time) + (time.Second * 10)
+		if sleepDuration > 0 {
+			glog.Infof("--Rate Limiting-- Tokens reached minimum reserve %d. Sleeping until reset in %v.\n", c.tokenReserve, sleepDuration)
+			time.Sleep(sleepDuration)
+		}
+	}
+}
+
+// withRetryAndRateLimit handles rate limiting and retry logic for a github API call.
+func (c *Client) withRetryAndRateLimit(call func() (*github.Response, error), action string) error {
+	var err error
+
+	for retryCount := 0; retryCount <= c.retries; retryCount++ {
+		var resp *github.Response
+		if resp, err = call(); err == nil {
+			c.limitRate(&resp.Rate)
+			return nil
+		} else if rlErr, ok := err.(*github.RateLimitError); ok {
+			c.limitRate(&rlErr.Rate)
+		} else if tfaErr, ok := err.(*github.TwoFactorAuthError); ok {
+			return tfaErr
+		}
+
+		if retryCount == c.retries {
+			return err
+		} else {
+			glog.Errorf("error %s: %v. Will retry.\n", action, err)
+			c.retrySleep(retryCount)
+		}
+	}
+	return err
+}
+
+// CreateStatus creates or updates a status context on the indicated reference.
+// This function limits rate and does retries if needed.
+func (c *Client) CreateStatus(owner, repo, ref string, status *github.RepoStatus) (*github.RepoStatus, *github.Response, error) {
+	glog.Infof("CreateStatus for ref '%s': %s:%s (%s)\n", ref, *status.Context, *status.State, *status.Description)
+	if c.dryRun {
+		return nil, nil, nil
+	}
+	var result *github.RepoStatus
+	var resp *github.Response
+	call := func() (*github.Response, error) {
+		var err error
+		result, resp, err = c.repoService.CreateStatus(owner, repo, ref, status)
+		return resp, err
+	}
+	err := c.withRetryAndRateLimit(call, fmt.Sprintf("creating status for ref '%s'", ref))
+	return result, resp, err
+}
+
+// GetCombinedStatus retrieves the CominedStatus for the specified reference.
+// This function limits rate, does retries if needed and handles pagination.
+func (c *Client) GetCombinedStatus(owner, repo, ref string) (*github.CombinedStatus, *github.Response, error) {
+	var status, result *github.CombinedStatus
+	var resp *github.Response
+
+	listOpts := &github.ListOptions{Page: 1, PerPage: 100}
+	lastPage := 1
+
+	call := func() (*github.Response, error) {
+		var err error
+		status, resp, err = c.repoService.GetCombinedStatus(owner, repo, ref, listOpts)
+		return resp, err
+	}
+	action := fmt.Sprintf("getting combined status for ref '%s'", ref)
+
+	for ; listOpts.Page <= lastPage; listOpts.Page++ {
+		if err := c.withRetryAndRateLimit(call, action); err != nil {
+			return result, resp, err
+		}
+		lastPage = resp.LastPage
+		if result == nil {
+			result = status
+		} else {
+			result.Statuses = append(result.Statuses, status.Statuses...)
```

Comment:
If that does occur it shouldn't matter since everything is synchronous and all the modes are idempotent.